tfjs-backend-webgpu/src/kernels/FromPixels_utils/from_pixels_webgpu.ts

```diff
@@ -23,38 +23,157 @@ export class FromPixelsProgram implements WebGPUProgram {
   dispatchLayout: {x: number[]};
+  variableNames: string[] = [];
   dispatch: [number, number, number];
+  workGroupSize: [number, number, number] = [256, 1, 1]; // The empirical value.
+
+  pipeline: GPUComputePipeline;
+  bindGroupLayout: GPUBindGroupLayout;
+
+  uniform: GPUBuffer;
+  lastUniformData = [0, 0];
+
+  inputTexture: GPUTexture = null;
+  lastPixelSize = {width: 0, height: 0};
+
+  private disposed = false;
 
   constructor(outputShape: number[]) {
     this.outputShape = outputShape;
+    this.workPerThread = outputShape[2];  // numChannels in outputShape.
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
-    this.dispatch = computeDispatch(this.dispatchLayout, this.outputShape);
+    this.dispatch = computeDispatch(
+      this.dispatchLayout, this.outputShape, this.workGroupSize,
+      [this.workPerThread, 1, 1]);
 
     this.userCode = `
-      void main() {
-        ivec3 coords = getOutputCoords();
-        int texR = coords[0];
-        int texC = coords[1];
-        int depth = coords[2];
-        vec2 uv = (vec2(texC, texR) + halfCR) / vec2(outShape.yx);
-
-        vec4 values = texelFetch(A, uv);
-        float value;
-        if (depth == 0) {
-          value = values.r;
-        } else if (depth == 1) {
-          value = values.g;
-        } else if (depth == 2) {
-          value = values.b;
-        } else if (depth == 3) {
-          value = values.a;
-        }
+    layout (local_size_x = ${this.workGroupSize[0]},
+      local_size_y = 1,
+      local_size_z = 1) in;
+    layout(set = 0, binding = 1, rgba8) uniform readonly image2D srcImage;
+    layout(set = 0, binding = 2) uniform Meta {
+      int size;
+      int numChannels;
+    } outShape;
 
-        setOutput(floor(value * 255.0 + 0.5));
+    void main() {
+      int flatIndexBase = int(gl_GlobalInvocationID.x) * outShape.numChannels;
+      ivec3 coords = getCoordsFromFlatIndex(flatIndexBase);
+      int texR = coords[0];
+      int texC = coords[1];
+      int depth = coords[2];
+      vec4 values = imageLoad(srcImage, ivec2(texC, texR));
+      for(int i = 0; i < outShape.numChannels; i++) {
+        float value = values[i];
+        int flatIndex = flatIndexBase + i;
+        if (flatIndex < outShape.size) {
+          result[flatIndex] = int(floor(255.0 * value));
+        }
       }
+    }
     `;
-    this.shaderKey = 'fromPixel';
+    this.shaderKey = `fromPixel${this.workGroupSize}`;
   }
+
+  setWebGPUBinary(
+    bindGroupLayout: GPUBindGroupLayout, pipeline: GPUComputePipeline) {
+      this.bindGroupLayout = bindGroupLayout;
+      this.pipeline = pipeline;
+    }
+
+    setUniform(device: GPUDevice, uniformData: number[]) {
+      // No need to update uniform buffer if no changes.
+      if (!uniformData ||
+        (uniformData[0] === this.lastUniformData[0] &&
+         uniformData[1] === this.lastUniformData[1])) {
+        return;
+      }
+
+      // TODO(tfjs-optimization): using new mapAsync API to update the buffer.
```

Comment:
I suppose this comment should be added in line 99. Just curious, why do you think mapAsync is better than writeBuffer?